src/array.rs

```diff
@@ -429,39 +441,42 @@ impl<T: TypeNum, D: Dimension> PyArray<T, D> {
         }
     }
 
-    /// Get the immutable view of the internal data of `PyArray`, as slice.
+    /// Returns the immutable view of the internal data of `PyArray` as slice.
+    /// Please consider the use of [`PyReadOnlyArray::as_slice`](../struct.PyReadOnlyArray.html)
+    /// instead of this.
     ///
-    /// Returns `ErrorKind::NotContiguous` if the internal array is not contiguous.
-    /// # Example
-    /// ```
-    /// use numpy::{PyArray, PyArray1};
-    /// use pyo3::types::IntoPyDict;
-    /// let gil = pyo3::Python::acquire_gil();
-    /// let py = gil.python();
-    /// let py_array = PyArray::arange(py, 0, 4, 1).reshape([2, 2]).unwrap();
-    /// assert_eq!(py_array.as_slice().unwrap(), &[0, 1, 2, 3]);
-    /// let locals = [("np", numpy::get_array_module(py).unwrap())].into_py_dict(py);
-    /// let not_contiguous: &PyArray1<i32> = py
-    ///     .eval("np.arange(10)[::2]", Some(locals), None)
-    ///     .unwrap()
-    ///     .downcast()
-    ///     .unwrap();
-    /// assert!(not_contiguous.as_slice().is_err());
-    /// ```
-    pub fn as_slice(&self) -> Result<&[T], NotContiguousError> {
+    /// # Safety
+    /// If the internal array is not readonly and can be mutated from Python code,
+    /// holding the slice might cause an undefined undefined behavior.
+    pub unsafe fn as_slice(&self) -> Result<&[T], NotContiguousError> {
         if !self.is_contiguous() {
             Err(NotContiguousError)
         } else {
-            Ok(unsafe { slice::from_raw_parts(self.data(), self.len()) })
+            Ok(slice::from_raw_parts(self.data(), self.len()))
         }
     }
 
-    /// Get the mmutable view of the internal data of `PyArray`, as slice.
-    pub fn as_slice_mut(&self) -> Result<&mut [T], NotContiguousError> {
+    /// Returns the view of the internal data of `PyArray` as `&[Cell<T>]`.
+    pub fn as_cell_slice(&self) -> Result<&[Cell<T>], NotContiguousError> {
         if !self.is_contiguous() {
             Err(NotContiguousError)
         } else {
-            Ok(unsafe { slice::from_raw_parts_mut(self.data(), self.len()) })
+            Ok(unsafe { slice::from_raw_parts(self.data() as _, self.len()) })
+        }
+    }
+
+    /// Returns the view of the internal data of `PyArray` as mutable slice.
+    ///
+    /// # Safety
+    /// If another reference to the internal data exists(e.g., `&[T]` or `ArrayView`),
+    /// it might cause an undefined undefined behavior.
+    ///
+    /// In such case, please consider the use of [`as_cell_slice`](#method.as_cell_slice),
+    pub unsafe fn as_slice_mut(&self) -> Result<&mut [T], NotContiguousError> {
```

Comment:
While you're breaking things here, might be worth calling this `as_mut_slice`, which is the same name std uses (e.g. `Vec::as_mut_slice`).

```suggestion
    pub unsafe fn as_mut_slice(&self) -> Result<&mut [T], NotContiguousError> {
```

Reply:
Yeah, but we follow the ndarray's `as_slice_mut` here.